os/net/ipv6/multicast/mpl.c

```diff
@@ -0,0 +1,1807 @@
+          /* We can use an S=0 Seed ID */
+          SEED_INFO_SET_LEN(locsiptr, 0);
+          break;
+        } /* Else fall down into the S = 3 case */
+        case 3:
+        seed_id_host_to_net(&((struct seed_info_s3*)locsiptr)->seed_id, &locssptr->seed_id);
+        SEED_INFO_SET_S(locsiptr, 3);
+        break;
+        case 1:
+        seed_id_host_to_net(&((struct seed_info_s1*)locsiptr)->seed_id, &locssptr->seed_id);
+        SEED_INFO_SET_S(locsiptr, 1);
+        break;
+        case 2:
+        seed_id_host_to_net(&((struct seed_info_s2*)locsiptr)->seed_id, &locssptr->seed_id);
+        SEED_INFO_SET_S(locsiptr, 2);
+        break;
+      }
+
+      /* Populate the seed info message vector */
+      memset(vector, 0, sizeof(uint8_t) * 32);
+      vec_len = 0;
+      locmmptr = locssptr->min_seq;
+      while(locmmptr != NULL) {
+        if(locmmptr->seq == SEQ_VAL_ADD(locssptr->min_seqno, vec_len)) {
+          BIT_VECTOR_SET_BIT(vector, vec_len);
+          locmmptr = locmmptr->next;
+        }
+        vec_len++;
+      }
+
+      /* Convert vector length from bits to bytes */
+      vec_size = (vec_len - 1) / 8 + 1;
+
+      SEED_INFO_SET_LEN(locsiptr, vec_size);
+
+      LOG_DBG("--- Control Message Entry ---\n");
+      LOG_DBG("Seed ID: ");
+      LOG_DBG_SEED(locssptr->seed_id);
+      LOG_DBG_("\n");
+      LOG_DBG("S=%u\n", locssptr->seed_id.s);
+      LOG_DBG("Min Sequence Number: %u\n", locssptr->min_seqno);
+      LOG_DBG("Size of message set: %u\n", vec_len);
+      LOG_DBG("Vector is %u bytes\n", vec_size);
+
+      /* Copy vector into payload and point ptr to next location */
+      switch(SEED_INFO_GET_S(locsiptr)) {
+      case 0:
+        memcpy(((void *)locsiptr) + sizeof(struct seed_info), vector, sizeof(uint8_t) * vec_size);
+        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info) + (sizeof(uint8_t) * vec_size);
+        payload_len += sizeof(struct seed_info) + (sizeof(uint8_t) * vec_size);
+        break;
+      case 1:
+        memcpy(((void *)locsiptr) + sizeof(struct seed_info_s1), vector, sizeof(uint8_t) * vec_size);
+        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info_s1) + (sizeof(uint8_t) * vec_size);
+        payload_len += sizeof(struct seed_info_s1) + (sizeof(uint8_t) * vec_size);
+        break;
+      case 2:
+        memcpy(((void *)locsiptr) + sizeof(struct seed_info_s2), vector, sizeof(uint8_t) * vec_size);
+        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info_s2) + (sizeof(uint8_t) * vec_size);
+        payload_len += sizeof(struct seed_info_s2) + (sizeof(uint8_t) * vec_size);
+        break;
+      case 3:
+        memcpy(((void *)locsiptr) + sizeof(struct seed_info_s3), vector, sizeof(uint8_t) * vec_size);
+        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info_s3) + (sizeof(uint8_t) * vec_size);
+        payload_len += sizeof(struct seed_info_s3) + (sizeof(uint8_t) * vec_size);
+        break;
+      }
+    }
```

Comment:
looks like a lot of redundancy in the code above, both within `case` clauses (e.g. `sizeof(uint8_t) * vec_size` etc.) and among clauses (all closes do more or less the same thing, only with different offsets and sizes)

Reply:
> looks like a lot of redundancy in the code above, within case clauses

Ah yes I see. I can sort that.

>and among clauses (all closes do more or less the same thing, only with different offsets and sizes)

How would you suggest tackling this? I can move most of the `+ vec_size` parts out of the cases, however I feel that affects the readability as the pointer arithmetic is then divided in two and it's less obvious as to it's intention.

Reply:
You're probably right, do what you think is best and I'll take another look!

Reply:
I've left it like this.

```c
switch(SEED_INFO_GET_S(locsiptr)) {
      case 0:
        memcpy(((void *)locsiptr) + sizeof(struct seed_info), vector, vec_size);
        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info) + vec_size;
        payload_len += sizeof(struct seed_info) + vec_size;
        break;
      case 1:
        memcpy(((void *)locsiptr) + sizeof(struct seed_info_s1), vector, vec_size);
        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info_s1) + vec_size;
        payload_len += sizeof(struct seed_info_s1) + vec_size;
        break;
      case 2:
        memcpy(((void *)locsiptr) + sizeof(struct seed_info_s2), vector, vec_size);
        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info_s2) + vec_size;
        payload_len += sizeof(struct seed_info_s2) + vec_size;
        break;
      case 3:
        memcpy(((void *)locsiptr) + sizeof(struct seed_info_s3), vector, vec_size);
        locsiptr = ((void *)locsiptr) + sizeof(struct seed_info_s3) + vec_size;
        payload_len += sizeof(struct seed_info_s3) + vec_size;
        break;
}
```
---
Do you think it would be better to move the `+ vec_size` parts from the 2nd and 3rd lines in each case into their own lines at the bottom (like below), or do you think I should keep it as it is?

```c
 locsiptr = ((void*)locsiptr) + vec_size;
 payload_len += vec_size;
```

Reply:
I am perhaps missing something, but isn't the only difference the size of the seed info structure? Why not store this in a variable, say `seed_info_len` in the four cases and then

```C
    memcpy(((void *)locsiptr) + seed_info_len, vector, vec_size);
    locsiptr = ((void *)locsiptr) + seed_info_len + vec_size;
    payload_len += seed_info_len + vec_size;
```

Could that work?

Reply:
No that works and I like that solution. Thanks!